table/tables/partition.go

```diff
@@ -714,9 +772,46 @@ func (lp *ForListPruning) locateListColumnsPartitionByRow(ctx sessionctx.Context
 	return location[0].PartIdx, nil
 }
 
-// buildListPartitionValueMap builds list columns partition value map for the specified column.
+// LocateRange locates partition range by the column range
+func (lp *ForListPruning) LocateRange(ctx sessionctx.Context, r *ranger.Range) ([]int, error) {
+	if r.LowVal[0].Kind() == types.KindMinNotNull {
+		r.LowVal[0] = types.GetMinValue(lp.PruneExpr.GetType())
+	}
+	if r.HighVal[0].Kind() == types.KindMaxValue {
+		r.HighVal[0] = types.GetMaxValue(lp.PruneExpr.GetType())
+	}
+	lowVal, isLowNull, err := lp.PruneExpr.EvalInt(ctx, chunk.MutRowFromDatums(r.LowVal).ToRow())
+	if err != nil {
+		return nil, err
+	}
+	highVal, isHighNull, err := lp.PruneExpr.EvalInt(ctx, chunk.MutRowFromDatums(r.HighVal).ToRow())
```

Comment:
Need to constrain function type, just same like range partition.

Because,  `lowValue < col < highValue` , doean't means  `fn(lowValue) < fn(col) < fn(highValue)`.

Such as the partition expression is `0 - a`; and the query condition is `1 < a < 3`, 
`fn(1) < fn(a) < fn(3)` => `-1 < fn(a) < -3` is not right.

Reply:
I think this change will be a much big, beter to pull another PR to do range prune optimization for list partition, this PR can only for list columns partition.

Reply:
It seems that it is easier that this PR for list partition, and pull another PR to do range prune optimization for list partition. Because I can reset to the commit for list partition.

Reply:
I have remove the code which is related to list partition, PTAL